### PR TITLE
Gridproperties iterator

### PR DESCRIPTION
**Pull request by joakim-hove**
Basic support to iterate over `GridProperty<T>` instances from a `GridProperties<T>` collection. Slightly over engineered solution to avoid the `std::pair< >` of the `std::map<>::iterator` and also to esacpe from the `std::shared_ptr< >`

### CODE REVIEW

File: opm/parser/eclipse/EclipseState/Grid/GridProperties.hpp [hunk 1]
```diff
@@ -58,13 +58,22 @@ namespace Opm {
...
-
     class Eclipse3DProperties;
 
+    template <typename T> class GridPropertyIterator;
+
+    template <typename T>
+    using storageType = typename std::map<std::string , std::shared_ptr<GridProperty<T> > >;
+
+    template <typename T>
+    using  storageIteratorType = typename std::map<std::string , std::shared_ptr<GridProperty<T> > >::const_iterator;
```
Comment (jokva) [inline opm/parser/eclipse/EclipseState/Grid/GridProperties.hpp]: Why don't you leverage your own aliases?

`template< typename T >` using const_iterator = storage< T >::const_iterator`

**Comment by jokva** [comment]
Isn't it better to just kill `shared_ptr` as storage, rather than going through this?

File: opm/parser/eclipse/EclipseState/Grid/GridProperties.cpp [hunk 2]
```diff
@@ -204,6 +204,16 @@ namespace Opm {
 
 
     template< typename T >
+    void GridProperties<T>::insertKeyword(const SupportedKeywordInfo& supportedKeyword) const {
+        int nx = m_eclipseGrid.getNX();
+        int ny = m_eclipseGrid.getNY();
+        int nz = m_eclipseGrid.getNZ();
+
+        m_properties.emplace( std::make_pair(supportedKeyword.getKeywordName() , GridProperty<T>( nx, ny , nz , supportedKeyword )));
```
Comment (jokva) [inline opm/parser/eclipse/EclipseState/Grid/GridProperties.cpp]: If you're using `emplace` you don't have to go via pair (that's the point). Save yourself some trouble:

`emplace( keywordname, GridProperty< T >( nx, ny, nz, supported )  )`

You might even be able to omit GridProperty< T > by using `{}`.

File: opm/parser/eclipse/EclipseState/Grid/GridProperties.hpp [hunk 3]
```diff
@@ -124,6 +146,17 @@ namespace Opm {
         void handleCOPYREGRecord( const DeckRecord& record, const GridProperty<int>& regionProperty );
 
 
+        struct iterator : public storageIteratorType {
+            iterator( storageIteratorType iter ) : storageIteratorType( iter ) { }
```
Comment (jokva) [inline opm/parser/eclipse/EclipseState/Grid/GridProperties.hpp]: Is this even necessary? Anyway, `using storageIterator::storageIterator` accomplishes the same.

File: opm/parser/eclipse/EclipseState/Grid/GridProperties.hpp [hunk 4]
```diff
@@ -124,6 +146,17 @@ namespace Opm {
...
+
+            const GridProperty<T>& operator*( ) {
+                const auto& pair = storageIteratorType::operator*( );
+                return pair.second;
```
Comment (jokva) [inline opm/parser/eclipse/EclipseState/Grid/GridProperties.hpp]: `return this->storageIteratorType::operator*().second` ;)

**Comment by joakim-hove** [comment]
Damn - I'm proud!

**Comment by jokva** [comment]
This currently has a subtle bug: Properties are created twice *if* they're auto generated. I suppose `autogeneratedProperties` should be `reference_wrapper` or just plain pointer.

File: opm/parser/eclipse/EclipseState/Grid/GridProperties.hpp (hunk 4 above)
Comment (joakim-hove) [inline opm/parser/eclipse/EclipseState/Grid/GridProperties.hpp]: Well - readability is in the mind of the beholder; I prefer the one with the temporary.

**Comment by joakim-hove** [comment]
> This currently has a subtle bug: Properties are created twice if they're auto generated. I suppose autogeneratedProperties should be reference_wrapper or just plain pointer. 

I might misunderstand you here - but I think you are wrong; the member `m_autoGeneratedProperties`is *only* a `std::set<std::string` of keyword names - not actual keywords.

**Comment by joakim-hove** [comment]
> I might misunderstand you here

And in case I am in the wrong (heaven forbid!) - well spotted; I still haven't found it.

**Comment by jokva** [comment]
You're absolutely right and I'm a potato.

**Comment by joakim-hove** [comment]
Merge?

**Comment by jokva** [comment]
Let me have a quick second look and we're good to go.

**Pull request merged**